src/types/sgrStack.cpp

```diff
@@ -136,10 +139,55 @@ namespace Microsoft::Console::VirtualTerminal
             result.SetItalic(savedAttribute.IsItalic());
         }
 
-        // Underline = 4,
-        if (validParts.test(SgrSaveRestoreStackOptions::Underline))
+        // Underline = 4, DoublyUnderlined = 21,
+        const bool isUnderlinedValid = validParts.test(SgrSaveRestoreStackOptions::Underline);
+        const bool isDoublyUnderlinedValid = validParts.test(SgrSaveRestoreStackOptions::DoublyUnderlined);
+        if (isUnderlinedValid && isDoublyUnderlinedValid)
+        {
+            // all the styles are valid, we can simply apply the saved style.
+            result.SetUnderlineStyle(savedAttribute.GetUnderlineStyle());
+        }
+        else if (isUnderlinedValid)
+        {
+            const auto savedUl = savedAttribute.GetUnderlineStyle();
+            const auto isUnderlinedOn = savedUl != UnderlineStyle::NoUnderline &&
+                                        savedUl != UnderlineStyle::DoublyUnderlined;
+            if (isUnderlinedOn)
+            {
+                result.SetUnderlineStyle(savedUl);
+            }
+            else
+            {
+                // Turn off singly and extended styles, but if the current style is
+                // doubly underlined, no need to overwrite it. This mimics having
+                // two flags each for singly and doubly underlined, where flag for
+                // doubly underlined would be left 'on' even if we had turned off
+                // the singly underlined.
+                if (result.GetUnderlineStyle() != UnderlineStyle::DoublyUnderlined)
+                {
+                    result.SetUnderlineStyle(UnderlineStyle::NoUnderline);
+                }
+            }
+        }
+        else if (isDoublyUnderlinedValid)
         {
-            result.SetUnderlined(savedAttribute.IsUnderlined());
+            const auto isDoublyUnderlinedOn = savedAttribute.GetUnderlineStyle() == UnderlineStyle::DoublyUnderlined;
+            if (isDoublyUnderlinedOn)
+            {
+                result.SetUnderlineStyle(UnderlineStyle::DoublyUnderlined);
+            }
+            else
+            {
+                // Turn off doubly underlined, but if the current style is
+                // singly underlined (or extended style), no need to overwrite it.
+                // This mimics having two flags each for singly and doubly
+                // underlined, where flag for singly underlined would be left 'on'
+                // even if we had turned off the doubly underlined.
+                if (result.GetUnderlineStyle() == UnderlineStyle::DoublyUnderlined)
+                {
+                    result.SetUnderlineStyle(UnderlineStyle::NoUnderline);
+                }
+            }

```

Comment:
Even after a while I still don't get this code... Why isn't it something like this?
```cpp
if (validParts.any(SgrSaveRestoreStackOptions::Underline, SgrSaveRestoreStackOptions::DoublyUnderlined))
{
    const auto style = savedAttribute.GetUnderlineStyle();
    const auto valid = style == UnderlineStyle::DoublyUnderlined
        ? validParts.test(SgrSaveRestoreStackOptions::DoublyUnderlined)
        : validParts.test(SgrSaveRestoreStackOptions::Underline);
    result.SetUnderlined(valid ? style : UnderlineStyle::NoUnderline);
}
```

Reply:
This is what I thought when I wrote that:

If we were to have separate flags for the `Underlined` and `DoublyUnderlined`, and the underlined is the only valid attribute (out of `DoublyUnderlined` and `Underlined`), and the state of `Underlined` was 'off' in the `savedAttribute`, that means in the current attribute, we would've reset the Underlined flag, while keeping the DoublyUnderlined state.

Since we do not have two flags, resetting `Underlined` means `NoUnderlined`, but what if the current attribute, here result, is `DoublyUnderlined`? Can we leave `DoublyUnderlined` as is, because the Underlined is already "off"?
This is tested at L168. The L188 does the same for the case when `DoublyUnderlined` is the only valid one.